### PR TITLE
purescript: fix build compiling with ghc 8.4.4

### LINKED ISSUE #53597: purescript failing to build
## Issue description

```
CallStack (from HasCallStack):
  die', called at libraries/Cabal/Cabal/Distribution/Simple/Configure.hs:950:20 in Cabal-2.4.0.1:Distribution.Simple.Configure
  configureFinalizedPackage, called at libraries/Cabal/Cabal/Distribution/Simple/Configure.hs:460:12 in Cabal-2.4.0.1:Distribution.Simple.Configure
  configure, called at libraries/Cabal/Cabal/Distribution/Simple.hs:596:20 in Cabal-2.4.0.1:Distribution.Simple
  confHook, called at libraries/Cabal/Cabal/Distribution/Simple/UserHooks.hs:67:5 in Cabal-2.4.0.1:Distribution.Simple.UserHooks
  configureAction, called at libraries/Cabal/Cabal/Distribution/Simple.hs:178:19 in Cabal-2.4.0.1:Distribution.Simple
  defaultMainHelper, called at libraries/Cabal/Cabal/Distribution/Simple.hs:115:27 in Cabal-2.4.0.1:Distribution.Simple
  defaultMain, called at Setup.hs:6:8 in main:Main
Setup: Encountered missing dependencies:
aeson >=1.0 && <1.4, base >=4.8 && <4.12
```

### Steps to reproduce

I've tried `nix-build -A purescript -K` on `master`, `channels/nixpkgs-unstable` and `channels/nixos-18.09`

## Technical details

```
- system: `"x86_64-linux"`
 - host os: `Linux 4.14.91, NixOS, 18.09.1834.9d608a6f592 (Jellyfish)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.1.3`
 - channels(me): `"home-manager-18.09"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

### PR DESCRIPTION
**Pull request by jacereda**
Closes #53597.

###### Things done

Not sure this is the right thing though...

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [X] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by endgame** [comment]
I don't use purescript, but I'd really like to see that aeson fix get merged. Hopefully it happens soon.

**Comment by domenkozar** [comment]
@GrahamcOfBorg build purescript

**Comment by jacereda** [comment]
Sorry about the mess.
I rebased to current master because a fix for aeson already landed in master.

**Comment by ryantm** [comment]
Hi, @jacereda.

Thanks for the contribution. Can you update the comments to include references to upstream tickets/issues for these issues?

Thanks,
Ryan

**Comment by jacereda** [comment]
I'm trying to verify if cereal/psqueues are still a problem but I'm getting the build failure below. I found a similar problem here: https://github.com/NixOS/nixpkgs/pull/50920#issuecomment-441478643

```
building [html]: targets for 4 source files that are out of date
updating environment: 4 added, 0 changed, 0 removed
reading sources... [ 25%] index
reading sources... [ 50%] intro
reading sources... [ 75%] invoking

Extension error:
The 'ghc-flag' directive is already registered to domain std
reading sources... [100%] markup
clang-5.0: warning: argument unused during compilation: '-nopie' [-Wunused-command-line-argument]
make[1]: *** [docs/users_guide/ghc.mk:16: docs/users_guide/build-html/users_guide/index.html] Error 2
make[1]: *** Waiting for unfinished jobs....
clang-5.0: warning: argument unused during compilation: '-nopie' [-Wunused-command-line-argument]

Extension error:
The 'ghc-flag' directive is already registered to domain std

looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [ 25%] index
make[1]: *** [docs/users_guide/ghc.mk:28: docs/users_guide/build-man/ghc.1] Error 2
writing output... [ 50%] intro
writing output... [ 75%] invoking
clang-5.0: warning: argument unused during compilation: '-nopie' [-Wunused-command-line-argument]
<<ghc: 20003168 bytes, 11 GCs, 233684/298496 avg/max bytes residency (2 samples), 24M in use, 0.000 INIT (0.004 elapsed), 0.021 MUT (0.876 elapsed), 0.019 GC (0.028 elapsed) :ghc>>
writing output... [100%] markup

/private/tmp/501/nix-build-ghc-8.4.4.drv-0/ghc-8.4.4/utils/haddock/doc/invoking.rst:457: WARNING: unknown option: -cpp
generating indices... genindex
writing additional pages... search
copying static files... done
copying extra files... done
dumping search index in English (code: en) ... done
dumping object inventory... done
build succeeded, 1 warning.

The HTML pages are in .build-html.
cp -R utils/haddock/doc/.build-html utils/haddock/doc/haddock
make: *** [Makefile:127: all] Error 2
builder for '/nix/store/h7zflbah2x7n9zqcpfif3x3wzb12hqj2-ghc-8.4.4.drv' failed with exit code 2
cannot build derivation '/nix/store/fd4jy5xjyjaj02wv3y8g8bwvabbhd95k-purescript-0.12.2.drv': 1 dependencies couldn't be built

```

**Comment by jacereda** [comment]
A sequential build failed as well. Still waiting for the job to finish, but I suspect it's due to the latest sphinx update.

```
cat utils/runghc/runghc.wrapper                                     >> inplace/bin/runghc
chmod +x                                                     inplace/bin/runghc
/nix/store/7p3m7ahh8p5wib5s8f1hfl7jxixqra2w-python3.7-sphinx-1.8.3/bin/sphinx-build -b html -d docs/users_guide/.doctrees-html -D latex_paper_size=letter docs/users_guide docs/users_guide/build-html/users_guide
Running Sphinx v1.8.3

Extension error:
The 'ghc-flag' directive is already registered to domain std
make[1]: *** [docs/users_guide/ghc.mk:16: docs/users_guide/build-html/users_guide/index.html] Error 2
make: *** [Makefile:127: all] Error 2
builder for '/nix/store/h7zflbah2x7n9zqcpfif3x3wzb12hqj2-ghc-8.4.4.drv' failed with exit code 2
cannot build derivation '/nix/store/dhc7s25szby99slfgsf3jdr60zl9v2ka-purescript-0.12.2.drv': 1 dependencies couldn't be built
error: build of '/nix/store/dhc7s25szby99slfgsf3jdr60zl9v2ka-purescript-0.12.2.drv' failed
```

**Comment by marsam** [comment]
thanks for working on this. btw I think 102c6248fde649b12212debac4a1c5bb77231a6c fixes https://github.com/NixOS/nixpkgs/issues/55662

**Comment by domenkozar** [comment]
@GrahamcOfBorg build purescript

**Comment by jacereda** [comment]
OK, the fix was wrong, second attempt...

**Comment by jacereda** [comment]
I think it should be ready now.

**Comment by domenkozar** [comment]
@GrahamcOfBorg build purescript

**Comment by jacereda** [comment]
Was the problem just a timeout? How do you see the full logs?

**Comment by jacereda** [comment]
Loos like @peti disabled the sphinx docs, should I rebase and reenable or leave it disabled?

**Comment by domenkozar** [comment]
@GrahamcOfBorg build purescript

File: pkgs/development/haskell-modules/configuration-ghc-8.6.x.nix [hunk 1]
```diff
@@ -45,7 +45,6 @@ self: super: {
   unordered-containers = dontCheck super.unordered-containers;
 
   # Test suite does not compile.
-  cereal = dontCheck super.cereal;
```
Comment (peti) [inline pkgs/development/haskell-modules/configuration-ghc-8.6.x.nix]: This change is unrelated to purescript. It would be better to submit it in a separate PR, because this change is totally uncontroversial and it can be merged very quickly.

File: pkgs/development/haskell-modules/configuration-ghc-8.6.x.nix [hunk 2]
```diff
@@ -54,7 +53,6 @@ self: super: {
   hpc-coveralls = doJailbreak super.hpc-coveralls; # https://github.com/guillaume-nargeot/hpc-coveralls/issues/82
   http-api-data = doJailbreak super.http-api-data;
   persistent-sqlite = dontCheck super.persistent-sqlite;
-  psqueues = dontCheck super.psqueues;    # won't cope with QuickCheck 2.12.x
```
Comment (peti) [inline pkgs/development/haskell-modules/configuration-ghc-8.6.x.nix]: ditto

File: pkgs/development/haskell-modules/configuration-ghc-head.nix [hunk 3]
```diff
@@ -60,7 +60,6 @@ self: super: {
   hpc-coveralls = doJailbreak super.hpc-coveralls; # https://github.com/guillaume-nargeot/hpc-coveralls/issues/82
   http-api-data = doJailbreak super.http-api-data;
   persistent-sqlite = dontCheck super.persistent-sqlite;
-  psqueues = dontCheck super.psqueues;    # won't cope with QuickCheck 2.12.x
```
Comment (peti) [inline pkgs/development/haskell-modules/configuration-ghc-head.nix]: ditto

File: pkgs/top-level/haskell-packages.nix [hunk 4]
```diff
@@ -3,6 +3,18 @@
 }:
 
 let
+  # sphinx update to 1.8.x broke ghc versions < 8.6
+  sphinx_1_7_9 = buildPackages.python3Packages.sphinx.overridePythonAttrs (o:
```
Comment (peti) [inline pkgs/top-level/haskell-packages.nix]: Please add the old Sphinx version into `python-packages.nix`.

**Review by peti — CHANGES_REQUESTED** [review]

**Comment by peti** [comment]
> Loos like @peti disabled the sphinx docs, should I rebase and reenable or leave it disabled?

It's preferable to build the docs, obviously, so if your fix works properly then please re-base it. I have also added a couple of other comments in my review.

**Comment by jacereda** [comment]
https://github.com/NixOS/nixpkgs/pull/56136
https://github.com/NixOS/nixpkgs/pull/56137
https://github.com/NixOS/nixpkgs/pull/56138
https://github.com/NixOS/nixpkgs/pull/56139

File: pkgs/top-level/haskell-packages.nix (hunk 4 above)
Comment (jacereda) [reply pkgs/top-level/haskell-packages.nix]: Done.

**Review by peti — APPROVED** [review]

**Comment by peti** [comment]
Test builds will be running at https://hydra.nixos.org/jobset/nixpkgs/haskell-updates ... I'll merge to `master` once these come back successfully.

**Comment by peti** [comment]
This PR broke all ghc-8.2.2 and 8.4.4 builds: https://hydra.nixos.org/eval/1506062. I have reverted it from `haskell-updates`.

**Comment by jacereda** [comment]
Doh! Sorry, try again please.

**Comment by jacereda** [comment]
This is the new PR: https://github.com/NixOS/nixpkgs/pull/56162
